### PR TITLE
Refactoring + adding 'extracts' method for image prompt generation

### PR DESCRIPTION
**Pull request by andreykurenkov**
Hey Sharon! Here's the PR.

The extracts method is pretty straighforward, you can see what it does via the prompt txt in inputs and the extracts method in scripts/run.py

Here's what sections method produces:
2     "start": [
  3         "a photograph of a storm chasers truck in the middle of the desert, full of equipment, black and white, dark and moody, 8k, storm chasers, enrique mata",
  4         "a photograph of a tornado, with deep blue sky, white clouds and green trees, the tornado in the distance, far out of focus",
  5         "a photograph of a storm chaser with a video camera standing in the middle of a lightning storm, on a dirt road, illuminated by a large lightning bolt, shallow depth of field, so    ft focus, colorful, high quality, by andrew michalowski, thunderstorm, beautiful, cinematic, storm",
  6         "a photograph of a storm chaser running towards a tornado, dramatic lighting, dramatic sky, close up shot of the tornado, rotating, in the distance",
  7         "a photograph of a storm chaser being thrown by a tornado by wayne barlowe"
  8     ],
  9     "middle": [
 10         "a photograph of a storm chasers truck with a huge lightning bolt piercing the sky, mid shot, shallow depth of field, dramatic lighting, dramatic weather, storm chasers",
 11         "a photograph of a tornado funnel, golden hour warm lighting, storm clouds, grunge, stormy sky, apocalyptic, rustling grass, wind, trees, debris",
 12         "a photograph of a storm chaser with a video camera and a microphone, standing in a field during a tornado, capturing the sound of the storm, the sky is filled with lightning, it     is raining, light and dark, the storm is violent, the tornado is in the distance, the sky is green",
 13         "a photograph of a storm chaser running towards a tornado, dramatic lighting, dramatic sky, red and blue and purple lighting, dramatic clouds, dramatic rain, dramatic wind, drama    tic lightning, dramatic thunder, dramatic face, dramatic vibe, dramatic atmosphere, dramatic perspective, dramatic mood, dramatic sky, dramatic cloud,",
 14         "a photograph of a storm chaser being thrown by a tornado across a highway, a giant evil tornado, which has an evil face and is wearing a hat, a green evil tornado, a stylized to    rnado, a tornado that looks like a dragon, a tornado that looks like a monster, an evil tornado with"
 15     ],
 16     "end": [
 17         "a photograph of a storm chasers truck parked in the middle of a flat open field, with a cloud of dust behind it, 2016",
 18         "a photograph of a tornado, a hurricane, fire and brimstone, hell, a nuclear bomb mushroom cloud, 4k, ultra hd, high definition, high quality, sharp, detail, crisp, realistic, ph    oto, high dynamic range, high res, high resolution,",
 19         "a photograph of a storm chaser with a video camera strapped to the top of his car, approaching a tornado at night, in the middle of a lightning storm, light from the lightning i    lluminating his car and the tornado from behind, light from the lightning striking the road creating a glowing outline of his car and",
 20         "a photograph of a storm chaser running towards a tornado with a sword in his hand, while a second man is running after him with a pitchfork in his hand, amazing detail, nice lig    hting, concept art, scared, terrified, dramatic, 8k",
 21         "a photograph of a storm chaser being thrown by a tornado, dramatic lighting, dramatic clouds, dramatic lighting, storm, tornado, backlit, boom"
 22     ]
 23 }

vs extracts method:

{
    2     "EXTRACTS": [
    3 ┊       "A woman stands next to a man who is holding a video camera and an energy drink. She looks annoyed.",
    4 ┊       "A woman is driving a car and talking to a man in the passenger seat. She is holding a pair of goggles and a monitor.",
    5 ┊       "Two people, a man and a woman, preparing for a storm. The man is putting on gloves and a hood, and the woman is holding a drink.",
    6 ┊       "A man in a truck driving to intercept a monster supercell, laser-focused on the road ahead. The supercell forms a giant tornado.",
    7 ┊       "a woman driving a car while a man films a tornado, black and white photography",
    8 ┊       "A man running with a video camera in heavy weather while his partner screams at him to come back, black and white photography",
    9 ┊       "A photo of a real tornado, with the storm chaser in the foreground, about to be hit.",
   10 ┊       "A woman tries to save a man who is injured and unconscious during a meteorological emergency, she is on her knees next to him and the funnel of the tornado is close by.",
   11 ┊       "Marlie Ziele, an atmospheric scientist, is trapped with Reed under a funnel cloud. She's holding his camera up to her face, preparing to die.",
   12 ┊       "A woman is trapped under a tornado in Oklahoma with her colleague, who is lying next to her not moving. She's trying to keep her shit together. The funnel is so close now she               can see the individual elements: Wind. Water. Earth. Fire.",
   13 ┊       "A man standing over a woman who is lying on the ground with a broken leg, she is in pain and he is telling her she is a hero, black and white photography"
   14     ]
   15 }
   16 {

As you can see, the extracts method does not specify style in most cases, and in some cases is too specific ("Marlie Ziele"). The sections method has more style, but hallucinates weird details sometimes (eg nuclear clouds).  I guess for extracts  method the txt prompt can be fine-tuned to specify for prompt text and prompt style, or we can pass it through the secondary engineering step used in sections.

Either way, I figure it makes sense to merge this and then iterate.

### CODE REVIEW

**Comment by andreykurenkov** [comment]
PS I guess I did not do git stuff right - run_longsd.py is now scripts/run.py (not new files were created here, just moved around and renamed)

PSS I verified that the previous method ('sections') still works, and that the new ones works too. I could not get to the image generation stage (some complaint about CUDA) -- maybe it'd be better to set default to one GPU.

**Comment by andreykurenkov** [comment]
Here's what I am thinking for next iteration of extracts prompt: 

> ~
> Suggest an image to illustrate for the following extract from a short story:
> You notice things others don’t when you spend your entire life stuck in a tower. Rapunzel knew the best combination of hair products from all available in the realm. She knew the health of the crops by the color of the sunlit fields. She thought it curious that across the Queen’s valley, a stone tower stood apart from the rest of the village, similar to the one in which she was imprisoned. She used her hand mirror to flash light at the other tower.
> ---
> content: rapunzel in a tower, using her hand mirror to flash light at the other tower.
> style: watercolor
> ~
> Suggest an image to illustrate for the following extract from a short story:
> “My second wish is that you help me expand the Earth.” Zeke was excited about making the wish, but he was a little concerned about the magnitude of his request. “The Earth is getting larger,” the genie confirmed. “But you can’t see it yet.” “When will it spread?” Zeke asked. “How long will it take?” The genie tilted his head and said, “The Earth will be double its current size in ten months, twenty-three days, and eight hours.”
> ---
> content: the earth getting larger and a ring around the earth
> style: impressionist painting
> ~
> Suggest an image to illustrate for the following extract from a short story:
> I was in the lab the day that the greatest discovery in the history of science was made (though I acknowledge that I may be biased). Our lab put multiverse theory into practice to create controlled, self-contained micro-universes within our own. We programmed the physics inside and watched planets evolve over massive geologic time scales. 
> ---
> content: a crystal ball containing a micro-universe
> style: digital 3D art, high resolution, vivid
> ~

File: README.md [hunk 1]
```diff
@@ -51,10 +75,11 @@ Check out the files for the example text, **Three Little Pigs**.
 - `effective_prompts_fs.txt`: Effective "prompt-English" to use for few-shot translation from English GPT-3 prompts to prompt-English (1884 tokens).
 
 #### Multi-processing Multi-GPU Note
-Multi-processing is optimized for 2 Titan RTXs, with 24GB RAM each. Changing the number of GPUs to parallelize on is a simple edit in `run_longsd.sh`: just copy the first line and change CUDA_VISIBLE_DEVICES to the appropriate GPU id. 
+Multi-processing is optimized for 2 Titan RTXs, with 24GB RAM each. Changing the number of GPUs to parallelize on is a simple edit in `run_longsd.sh`: just copy the first line and change CUDA_VISIBLE_DEVICES to the appropriate GPU id.
```
Comment (sharonzhou) [inline README.md:78]: now `run.sh`?

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
+name: sba
```
Comment (sharonzhou) [inline environment.yml:1]: Change this back to slimmed down pip reqs. It works better for people using different venvs and is a lot cleaner.

File: inputs/extracts_prompt_prefix.txt [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Suggest descriptions of images to illustrate for the following paragraphs from short stories:
```
Comment (sharonzhou) [inline inputs/extracts_prompt_prefix.txt:1]: Wouldn't you only want to have this line here? The other text is story-specific that should be extracted automatically from a txt file.

File: inputs/sample_input.txt [hunk 4]
```diff
@@ -0,0 +1,81 @@
+Marlie nearly turned around when Reed stepped out of the souped-up armored truck, pounding an extra-large energy drink. This was not her scene, but she couldn't trust anyone else to deploy her drone system in the middle of the storm. If she wanted to finish her PhD thesis sometime in the next decade, she'd have to ride shotgun with this adrenaline-junkie and his video camera.
```
Comment (sharonzhou) [inline inputs/sample_input.txt:1]: Change sample to three_little_pigs.

And if keeping another one, make it the goats one.

File: scripts/gen_docx.py [hunk 5]
```diff
@@ -17,17 +16,17 @@ def dump_images_captions_docx(title, prompts_and_image_paths):
     title = title.split('.')[0]
     title = title.replace('_', ' ').title()
     ts = int(time.time())
-    document.save(f'docx/{title} - Illustrations - {ts}.docx')
+    document.save(f'outputs/docx/{title} - Illustrations - {ts}.docx')
```
Comment (sharonzhou) [inline scripts/gen_docx.py:19]: Hmm not sure if I'd agree w/ adding another level of directories

**Comment by sharonzhou** [comment]
Thanks for starting on this! Before merging, make a much more slimmed down PR without renaming and restructuring (let's discuss afterwards). And for extracts, just make it a flag in params, not a separate script.

**Comment by andreykurenkov** [comment]
oh, there is not separate script, it's just renamed and looks like an entirely new script (by accident). 
IMO it's less messy with restructuring (fewer top level files) but yeah can discuss in separate merge.

**Pull request closed without merging**